server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -266,6 +267,26 @@ public static String threadName(final String nodeName, final String namePrefix)
         return "elasticsearch" + (nodeName.isEmpty() ? "" : "[") + nodeName + (nodeName.isEmpty() ? "" : "]") + "[" + namePrefix + "]";
     }
 
+    // to be used in assertions only.
+    public static boolean differentExecutors(Thread thread1, Thread thread2) {
+        assert thread1 != thread2 : "only call for different threads";
+        String thread1Name = executorName(thread1);
+        String thread2Name = executorName(thread2);
+        return thread1Name == null || thread2Name == null || thread1Name.equals(thread2Name) == false;
+    }
+
+    // visible for tests
+    static String executorName(Thread thread) {
+        String name = thread.getName();
+        int executorNameEnd = name.lastIndexOf(']', name.length() - 2);
```

Comment:
nit A comment on why `name.length() - 2` would be helpful. Why is it like that? I understand that probably there's another `]` you'd like to skip? But I don't remember how the thread name is formatted out of the top of my head. Reading

```
    public static String threadName(final String nodeName, final String namePrefix) {
        // TODO missing node names should only be allowed in tests
        return "elasticsearch" + (nodeName.isEmpty() ? "" : "[") + nodeName + (nodeName.isEmpty() ? "" : "]") + "[" + namePrefix + "]";
    }
```

and assuming namePrefix is the executor name, I'm not sure where the other `]` would be.

Reply:
The thread naming is something like `XXXX[executor-name][T#NNN]`. We need to skip the running number part. The index passed into `lastIndexOf` is inclusive.

I added a short comment on this.

Reply:
Maybe we could use a regex here? maybe that's clearer? (famous last words)

Reply:
This time there's even _two_ obligatory XCKDs:
- https://xkcd.com/208/
- https://xkcd.com/1171/

Reply:
I prefer the manual parsing done here, seems simplest.